### PR TITLE
LocalDatastore can update from server

### PR DESCRIPTION
**Pull request by dplewis**
Allows for updating LocalDatastore with server data

Any advice would be helpful

### CODE REVIEW

**Comment by flovilmart** [comment]
Did you check how the other SDKs implémented those features? I’m not sure the server is properly equipped to ‘sync’ data when it’s been off sync? 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by dplewis** [comment]
> Did you check how the other SDKs implémented those features?

If by other SDK's you mean iOS and android LDS, I don't believe they have this kind of feature. They do have `saveEventually` for syncing

> I’m not sure the server is properly equipped to ‘sync’ data when it’s been off sync?

I'm not looking for a true sync feature with this PR. Just when network reconnect perform a query.

**Comment by flovilmart** [comment]
That’s not what I meant, when you will get network connection again, you will attempt to save object, and may override previous operations. But I guess you already knew that.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by flovilmart** [comment]
The REstController also is not the best spot to get this feature in, as it’s responsibility is solely to connect to the API. It has nothing to do with the LDS.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by dplewis** [comment]
> you will attempt to save object, and may override previous operations

This attempt doesn't happen on ie9 and when it does happen it only tries for 4 seconds max

> The REstController also is not the best spot to get this feature in

When do you believe the best implementation is? I would like to the SDK to handle it. Any advice would be appreciated

**Comment by dplewis** [comment]
@flovilmart How does this look?

**Comment by flovilmart** [comment]
@dplewis ask @acinader :) 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: src/Parse.js [hunk 1]
```diff
@@ -225,6 +225,23 @@ Parse.dumpLocalDatastore = function() {
     return Parse.LocalDatastore._getAllContents();
   }
 }
+/**
```
Comment (acinader) [inline src/Parse.js]: why here?  why not in LocalDatastore?

File: src/LocalDatastore.js [hunk 2]
```diff
@@ -259,6 +260,54 @@ const LocalDatastore = {
...
+      }
+    }
+    if (keys.length === 0) {
+      return;
+    }
+    this.isSyncing = true;
+    const pointersHash = {};
+    for (const key of keys) {
+      const [className, objectId] = key.split('_');
+      pointersHash[className] = pointersHash[className] || new Set();
```
[the remark this answers is not in the document]
Comment (dplewis) [reply src/LocalDatastore.js]: I modified this from [include()]( https://github.com/parse-community/parse-server/blob/master/src/RestQuery.js#L787)

File: src/Parse.js [hunk 3]
```diff
@@ -225,6 +225,23 @@ Parse.dumpLocalDatastore = function() {
...
+ * Updates Local Datastore from Server
+ *
+ * <pre>
+ * await Parse.updateLocalDatastoreFromServer();
+ * </pre>
+ *
+ * @static
+ */
+Parse.updateLocalDatastoreFromServer = function() {
+  if (!Parse.LocalDatastore.isEnabled) {
```
[the remark this answers is not in the document]
Comment (dplewis) [reply src/Parse.js]: Good question this should fail

File: src/LocalDatastore.js [hunk 4]
```diff
@@ -259,6 +260,65 @@ const LocalDatastore = {
...
+      const objects = [].concat.apply([], responses);
+      const pinPromises = objects.map((object) => {
+        const objectKey = this.getKeyForObject(object);
+        return this.pinWithName(objectKey, object._toFullJSON());
+      });
+      await Promise.all(pinPromises);
+      this.isSyncing = false;
+    } catch(error) {
+      console.log('Error syncing LocalDatastore'); // eslint-disable-line
+      console.log(error); // eslint-disable-line
```
Comment (acinader) [inline src/LocalDatastore.js]: any reason not to use parse's logger so this stuff shows up in aggregated logs?

see middlewares, MogoTransform, PromiseRouter, etc. for examples

Comment (dplewis) [reply src/LocalDatastore.js]: something like

```
const Logger = require('parse-server/lib/logger');

Logger.logger.error(`Error syncing LocalDatastore: ${error}`);
console.log('Error syncing LocalDatastore', error);
```

Comment (acinader) [reply src/LocalDatastore.js]: Any reason to not follow the exact pattern we have already?

i.e. 

https://github.com/parse-community/parse-server/blob/master/src/PromiseRouter.js#L10
```
import log from './logger';
...
log.error('message', object);
...
```

and remove the console.log and the eslint exception comment?

Comment (dplewis) [reply src/LocalDatastore.js]: Thats the parse-server repo, there isn't a logger in this repo.

Comment (acinader) [reply src/LocalDatastore.js]: ha.  riiiight.  yeah, lgtm. ;) . I don't use the client js sdk is the only lame excuse i have for that one.

**Pull request merged**